### ISSUE TITLE
Both generate.concurrency as generate.interval are flawed

### ISSUE THREAD

**Issue by pimlie** [nuxt/nuxt#4122]
### Version

[v2.2.0](https://github.com/nuxt.js/releases/tag/v2.2.0)

### Reproduction link

[https://pimlie.github.io/nuxt-generate-benchmarks/](https://pimlie.github.io/nuxt-generate-benchmarks/)

### Steps to reproduce

See the repro link for all the numbers.

### What is expected ?

See below

### What is actually happening?

See below

### Additional comments?

As these changes could be quite challenging, I though it would better to first make a Bug Report so we can discuss my benchmark results before I make a PR with the accepted proposal.

The current `generate.concurrency` and `generate.interval` implementations are both flawed. I have created some benchmarks which shows the behaviour and tested different new implementations.

With regards to concurrency, the main reason this is almost never noticed is because the default value is high and the flawed implementation is only noticeable when you set a low value or with really specific workload patterns. The current implementation is not really concurrent, its more batch'ed. A concurrent implementation means there are always `[num]` running jobs, but currently if you have a concurrency of 10 of which 1 page takes 1 sec to render and the other 9 takes 1ms, you still have to wait until that one page has finished. This isnt a problem if that page is fully blocking but when its waiting for something (eg an API request) you could have generated other pages in the mean time.

With regards to interval, the current implementation has no protection against differences in page workloads. Eg when you first have a long blocking page to render (which duration exceeds the set interval value) then when Node's event loop becomes available again (with the current implementation) the wait time for the next queued page (or pages) has immediately elapsed which could trigger multiple API requests within the same time block.

If you check the [`concurrency method`](https://pimlie.github.io/nuxt-generate-benchmarks/) benchmarks you can see the results for different proposals. These benchmarks were all done with `interval = 0`, some are also not _full featured_ but were added to compare overhead of (a part of) the implementation. 
If you have a different proposal I am happy to add it. The used implementations can be viewed [here](https://github.com/pimlie/nuxt.js/blob/feat-routes-generator/lib/builder/generator.js#L102). Based on these results it looked to me that `waitthrottled` would be the best implementation. 

Please also check the `interval throttling` benchmark, this is a comparison between how API calls are 'throttled' currently and with the proposed implementation. The benchmark was run with a special prepared workload to trigger the flaw in the old implementation.
The number shown on top of the bars is the `the count of interval blocks in which more than two pages where generated`. Maybe a bit cryptic, but when you set an interval you expect that within that interval only 1 page renders (due to measurement errors this could sometimes be 2, so therefor its >2). Ideally this should be 0 of course.

--edit--
`#1` and `#2` for the `concurrency method` benchmark just means I did two runs of the benchmark

<div align="right"><sub><em>This bug report is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c7978">#c7978</a>)</em></sub></div>

**Comment by Atinux** [comment]
Should be enchanced with the workers to handle concurrency inside the renderer.

We talked with @pi0 about this, and will handle a flag in the `ssrContext` directly instead of `process.static`.

**Comment by simplenotezy** [comment]
Thanks for your great efforts @pimlie into documenting this issue; it must've taken you a great time to prepare all this.

I wish I understood your benchmarks better, they'll need a bit more description IMO.

I'm also interested about proper concurrency/interval set ups, for maximum performance.

I experienced a gain, going from deploy time between 2-3000 seconds to around 600 seconds, after I lowered concurrency from the default 250 value, to 50.
